File: typescript/tests/receipts.test.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * Tests for cryptographic redaction receipts.
 */

import { describe, it, expect, beforeEach, vi } from 'vitest';
import { harden } from '../src/harden.js';
import { _resetHealthForTests, getHealthSnapshot } from '../src/health.js';
import { hmacSha256Hex, sha256Hex } from '../src/hash.js';
import { sanitizePayload, resetPiiRulesForTests, registerPiiRule } from '../src/pii.js';
import {
  canonicalJson,
  enableReceipts,
  getEmittedReceiptsForTests,
  emitReceipt,
  MissingReceiptSinkError,
  receiptPayload,
  TEST_RECEIPT_CAPACITY,
  TestReceiptCollector,
  resetReceiptsForTests,
  signReceipt,
  _setTestModeForTests,
} from '../src/receipts.js';
import type { RedactionReceipt } from '../src/receipts.js';

const enc = new TextEncoder();

function makeReceipt(): RedactionReceipt {
  return {
    receiptId: 'r-1',
    timestamp: '2026-08-07T00:00:00.000Z',
    serviceName: 'svc',
    fieldPath: 'user.password',
    action: 'redact',
    originalHash: 'a'.repeat(64),
    hmac: '',
  };
}

beforeEach(() => {
  resetPiiRulesForTests();
  resetReceiptsForTests();
});

describe('receipts disabled by default', () => {
  it('emits no receipts before enableReceipts is called', () => {
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    expect(getEmittedReceiptsForTests()).toHaveLength(0);
  });
});

describe('receipts emitted when enabled', () => {
  it('generates a receipt when a sensitive field is sanitized', () => {
    enableReceipts({ enabled: true, serviceName: 'test-svc' });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    const r = receipts[0];
    expect(r.fieldPath).toBe('password');
    expect(r.action).toBe('redact');
    expect(r.receiptId.length).toBeGreaterThan(0);
  });
});

describe('receipt original_hash is SHA-256 over the canonical JSON form', () => {
  it('hashes the JCS serialization, not the display form', () => {
    enableReceipts({ enabled: true });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    expect(receipts[0].originalHash).toBe(sha256Hex('"secret123"'));
  });

  it('distinguishes a numeric value from its string spelling', () => {
    // The point of canonicalizing rather than calling String(): hashing the
    // display form makes the number 1 and the string "1" indistinguishable in
    // the audit trail.
    enableReceipts({ enabled: true });
    const asNumber = { password: 1 };
    const asString = { password: '1' };
    sanitizePayload(asNumber);
    sanitizePayload(asString);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(2);
    expect(receipts[0].originalHash).not.toBe(receipts[1].originalHash);
    expect(receipts[0].originalHash).toBe(sha256Hex('1'));
    expect(receipts[1].originalHash).toBe(sha256Hex('"1"'));
  });
});

describe('receipt HMAC when key provided', () => {
  it('computes HMAC correctly with a signing key', () => {
    enableReceipts({ enabled: true, signingKey: 'test-key' });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    const r = receipts[0];
    expect(r.hmac).not.toBe('');
    const payload = `${r.receiptId}|${r.timestamp}|${r.fieldPath}|${r.action}|${r.originalHash}`;
    expect(receiptPayload(r)).toBe(payload);
    expect(r.hmac).toBe(hmacSha256Hex(enc.encode('test-key'), enc.encode(payload)));
  });

  it('is a real HMAC, not a keyed digest of the concatenation', () => {
    // The predecessor computed sha256("key|payload"), which reproduces no
    // cross-language vector and is length-extendable.
    enableReceipts({ enabled: true, signingKey: 'test-key' });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const r = getEmittedReceiptsForTests()[0];
    expect(r.hmac).not.toBe(sha256Hex(`test-key|${receiptPayload(r)}`));
  });
});

describe('receipt HMAC empty when no key', () => {
  it('leaves hmac as empty string when no signing key', () => {
    enableReceipts({ enabled: true });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    expect(receipts[0].hmac).toBe('');
  });
});

describe('receipt tamper detection', () => {
  it('produces a different HMAC after tampering with field_path', () => {
    enableReceipts({ enabled: true, signingKey: 'test-key' });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    const r = receipts[0];
    const tamperedPayload = `${r.receiptId}|${r.timestamp}|tampered.path|${r.action}|${r.originalHash}`;
    const tamperedHMAC = hmacSha256Hex(enc.encode('test-key'), enc.encode(tamperedPayload));
    expect(r.hmac).not.toBe(tamperedHMAC);
  });
});

describe('enable/disable toggle', () => {
  it('unregisters the hook when enabled=false', () => {
    enableReceipts({ enabled: true });
    enableReceipts({ enabled: false });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    expect(getEmittedReceiptsForTests()).toHaveLength(0);
  });
});

describe('receipt_id is UUID format', () => {
  it('receipt_id has UUID format (36 chars, dashes)', () => {
    enableReceipts({ enabled: true });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    const rid = receipts[0].receiptId;
    expect(rid).toHaveLength(36);
    expect(rid[8]).toBe('-');
    expect(rid[13]).toBe('-');
    expect(rid[18]).toBe('-');
    expect(rid[23]).toBe('-');
  });
});

describe('service name is set in receipt', () => {
  it('receipt.serviceName reflects the configured service name', () => {
    enableReceipts({ enabled: true, serviceName: 'my-service' });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    expect(receipts[0].serviceName).toBe('my-service');
  });

  it('receipt.serviceName defaults to "unknown" when not specified', () => {
    enableReceipts({ enabled: true });
    const obj = { password: 'secret123' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    expect(receipts[0].serviceName).toBe('unknown');
  });
});

describe('production mode (testMode=false) requires a sink', () => {
  it('delivers to the configured sink instead of the test collector', () => {
    _setTestModeForTests(false);
    try {
      const delivered: unknown[] = [];
      enableReceipts({
        enabled: true,
        sink: {
          emit: (r) => {
            delivered.push(r);
            return true;
          },
        },
      });
      const obj = { password: 'secret123' }; // pragma: allowlist secret
      sanitizePayload(obj);
      expect(delivered).toHaveLength(1);
      expect(getEmittedReceiptsForTests()).toHaveLength(0);
    } finally {
      _setTestModeForTests(true);
    }
  });

  it('refuses to enable without a sink rather than discarding receipts', () => {
    // Previously this configuration signed a receipt for every redaction and
    // then dropped it, so a service could believe it had an audit trail.
    _setTestModeForTests(false);
    try {
      expect(() => {
        enableReceipts({ enabled: true });
      }).toThrow(MissingReceiptSinkError);
      // The message is the whole value of this error: it has to say what was
      // misconfigured and what to do, or a caller just sees a thrown class.
      let caught: unknown;
      try {
        enableReceipts({ enabled: true });
      } catch (error) {
        caught = error;
      }
      expect((caught as Error).name).toBe('MissingReceiptSinkError');
      expect((caught as Error).message).toBe(
        'receipts are enabled but no receiptSink is configured; ' +
          'generated receipts would be computed and then discarded. ' +
          'Pass a ReceiptSink, or disable receipts.',
      );
    } finally {
      _setTestModeForTests(true);
    }
  });

  it('still allows disabling receipts without a sink', () => {
    _setTestModeForTests(false);
    try {
      expect(() => {
        enableReceipts({ enabled: false });
      }).not.toThrow();
    } finally {
      _setTestModeForTests(true);
    }
  });
});

describe('signReceipt defaults', () => {
  it('defaults serviceName to "unknown" when the caller omits it', () => {
    const receipt = signReceipt('value', {
      receiptId: 'r-1',
      timestamp: '2026-08-07T00:00:00.000Z',
      fieldPath: 'user.password',
      action: 'redact',
    });
    expect(receipt.serviceName).toBe('unknown');
  });

  it('uses the caller service name when given', () => {
    const receipt = signReceipt('value', {
      receiptId: 'r-1',
      timestamp: '2026-08-07T00:00:00.000Z',
      fieldPath: 'user.password',
      action: 'redact',
      serviceName: 'billing',
    });
    expect(receipt.serviceName).toBe('billing');
  });

  it('hashes non-finite numbers as the null spelling, so digests agree across SDKs', () => {
    const opts = {
      receiptId: 'r-1',
      timestamp: '2026-08-10T00:00:00.000Z',
      fieldPath: 'metrics.ratio',
      action: 'redact',
    };
    const nullHash = signReceipt(null, opts).originalHash;
    expect(signReceipt(Number.NaN, opts).originalHash).toBe(nullHash);
    expect(signReceipt(Number.POSITIVE_INFINITY, opts).originalHash).toBe(nullHash);
    expect(signReceipt(Number.NEGATIVE_INFINITY, opts).originalHash).toBe(nullHash);
    // Hardening passes non-finite numbers through unchanged (it no longer
    // redacts them to '***'), so a hardened value signs to the same digest —
    // in every SDK, not just this one.
    expect(signReceipt(harden(Number.NaN), opts).originalHash).toBe(nullHash);
  });
});

describe('canonicalJson', () => {
  it('orders object keys by UTF-16 code unit, not insertion order', () => {
    expect(canonicalJson({ zeta: 1, Alpha: 2, alpha: 3 })).toBe('{"Alpha":2,"alpha":3,"zeta":1}');
  });

  it('collapses negative zero, which JSON.stringify already does and JCS requires', () => {
    expect(canonicalJson({ balance: -0 })).toBe('{"balance":0}');
  });

  it('normalizes values JSON cannot encode to null', () => {
    expect(canonicalJson(Number.NaN)).toBe('null');
    expect(canonicalJson(Number.POSITIVE_INFINITY)).toBe('null');
    expect(canonicalJson(Number.NEGATIVE_INFINITY)).toBe('null');
    expect(canonicalJson(undefined)).toBe('null');
  });

  it('normalizes bigint, symbol, and function without throwing', () => {
    // JSON.stringify throws a TypeError on a bigint. Throwing here would turn
    // a log call into an exception, because canonicalization runs inside the
    // redaction hook.
    expect(canonicalJson(10n)).toBe('null');
    expect(canonicalJson(Symbol('s'))).toBe('null');
    expect(canonicalJson(() => 1)).toBe('null');
    expect(canonicalJson({ n: 10n })).toBe('{"n":null}');
  });

  it('encodes the JSON scalars it can', () => {
    expect(canonicalJson('a"b')).toBe('"a\\"b"');
    expect(canonicalJson(1.5)).toBe('1.5');
    expect(canonicalJson(true)).toBe('true');
    expect(canonicalJson(null)).toBe('null');
  });

  it('keeps undefined array elements and object values as null rather than dropping them', () => {
    // A vanishing key would change the hashed structure.
    expect(canonicalJson([1, undefined, 3])).toBe('[1,null,3]');
    expect(canonicalJson({ a: undefined })).toBe('{"a":null}');
  });

  it('recurses through nested arrays and objects', () => {
    expect(canonicalJson({ b: [{ d: 1, c: 2 }], a: null })).toBe('{"a":null,"b":[{"c":2,"d":1}]}');
  });

  it('serializes a self-referential object as null instead of recursing to a RangeError', () => {
    // Mirrors Python receipts._canonical: a composite already on the current
    // serialization path spells null, the same spelling every other
    // JSON-unencodable value gets.
    const cyclic: Record<string, unknown> = { name: 'root' };
    cyclic['self'] = cyclic;
    expect(canonicalJson(cyclic)).toBe('{"name":"root","self":null}');
  });

  it('serializes an array cycle as null', () => {
    const arr: unknown[] = [1];
    arr.push(arr);
    expect(canonicalJson(arr)).toBe('[1,null]');
  });

  it('serializes an indirect cycle at the point it closes', () => {
    const inner: Record<string, unknown> = {};
    const outer = { inner };
    inner['back'] = outer;
    expect(canonicalJson(outer)).toBe('{"inner":{"back":null}}');
  });

  it('serializes a shared acyclic subtree fully at every occurrence', () => {
    // The guard is path-scoped: a value leaves the set when its subtree
    // completes, so sharing is not mistaken for a cycle.
    const shared = { k: 1 };
    expect(canonicalJson({ a: shared, b: shared })).toBe('{"a":{"k":1},"b":{"k":1}}');
    expect(canonicalJson([shared, shared])).toBe('[{"k":1},{"k":1}]');
  });
});

describe('receipts from custom PII rules (covers _applyRuleFull receipt hook path)', () => {
  it('emits a receipt when a custom rule matches a field', () => {
    enableReceipts({ enabled: true, serviceName: 'rule-svc' });
    registerPiiRule({ path: 'user.email', mode: 'redact' });
    const obj = { user: { email: 'alice@example.com' } };
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    expect(receipts[0].fieldPath).toBe('user.email');
    expect(receipts[0].action).toBe('redact');
  });
});

describe('receipts emitted for secret detection in non-blocked keys (covers _applyDefaultSensitiveKeyRedaction secret path)', () => {
  it('emits a receipt when a secret pattern is detected in an unblocked field', () => {
    enableReceipts({ enabled: true });
    // 'custom_field' is not a blocked key — secret is detected by VALUE pattern (40+ hex chars)
    const obj = { custom_field: 'a1b2c3d4e5f6a1b2c3d4e5f6a1b2c3d4e5f6a1b2' }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts).toHaveLength(1);
    expect(receipts[0].fieldPath).toBe('custom_field');
  });
});

describe('receipts emitted for sensitive keys inside arrays (covers array branch in _applyDefaultSensitiveKeyRedaction)', () => {
  it('emits receipts for sensitive fields inside array items', () => {
    enableReceipts({ enabled: true });
    const obj = { users: [{ password: 'pass1' }, { password: 'pass2' }] }; // pragma: allowlist secret
    sanitizePayload(obj);
    const receipts = getEmittedReceiptsForTests();
    expect(receipts.length).toBeGreaterThanOrEqual(2);
    expect(receipts.every((r) => r.action === 'redact')).toBe(true);
  });
});

describe('sink accounting (health.receiptFailures)', () => {
  beforeEach(() => {
    _resetHealthForTests();
  });

  it('counts a refused receipt without logging', () => {
    // The error path must never log: the logger produces redactions,
    // redactions produce receipts, and a sink that fails on every receipt
    // would otherwise drive an unbounded log -> receipt -> log cycle.
    const consoleSpy = vi.spyOn(console, 'warn').mockImplementation(() => undefined);
    const errorSpy = vi.spyOn(console, 'error').mockImplementation(() => undefined);
    try {
      emitReceipt(makeReceipt(), { emit: () => false });
      expect(getHealthSnapshot().receiptFailures).toBe(1);
      expect(consoleSpy).not.toHaveBeenCalled();
      expect(errorSpy).not.toHaveBeenCalled();
    } finally {
      consoleSpy.mockRestore();
      errorSpy.mockRestore();
    }
  });

  it('counts a throwing sink the same as a refusing one', () => {
    emitReceipt(makeReceipt(), {
      emit: () => {
        throw new Error('sink is down');
      },
    });
    expect(getHealthSnapshot().receiptFailures).toBe(1);
  });

  it('does not count an accepted receipt', () => {
    const sink = new TestReceiptCollector();
    emitReceipt(makeReceipt(), sink);
    expect(getHealthSnapshot().receiptFailures).toBe(0);
    expect(sink.receipts).toHaveLength(1);
  });

  it('accumulates across failures', () => {
    const failing = { emit: () => false };
    emitReceipt(makeReceipt(), failing);
    emitReceipt(makeReceipt(), failing);
    expect(getHealthSnapshot().receiptFailures).toBe(2);
  });

  it('counts refusals raised through the live redaction path', () => {
    _setTestModeForTests(false);
    try {
      enableReceipts({ enabled: true, sink: { emit: () => false } });
      const obj = { password: 'secret123' }; // pragma: allowlist secret
      sanitizePayload(obj);
      expect(getHealthSnapshot().receiptFailures).toBe(1);
    } finally {
      _setTestModeForTests(true);
    }
  });
});

describe('TestReceiptCollector capacity', () => {
  it('retains the newest receipts once full, dropping the oldest', () => {
    const sink = new TestReceiptCollector();
    for (let i = 0; i < TEST_RECEIPT_CAPACITY + 5; i++) {
      sink.emit({ ...makeReceipt(), receiptId: `id-${String(i)}` });
    }
    expect(sink.receipts).toHaveLength(TEST_RECEIPT_CAPACITY);
    expect(sink.receipts[0].receiptId).toBe('id-5');
    expect(sink.receipts[TEST_RECEIPT_CAPACITY - 1].receiptId).toBe(
      `id-${String(TEST_RECEIPT_CAPACITY + 4)}`,
    );
  });

  it('keeps every receipt below the cap', () => {
    const sink = new TestReceiptCollector();
    for (let i = 0; i < TEST_RECEIPT_CAPACITY; i++) {
      sink.emit({ ...makeReceipt(), receiptId: `id-${String(i)}` });
    }
    expect(sink.receipts).toHaveLength(TEST_RECEIPT_CAPACITY);
    expect(sink.receipts[0].receiptId).toBe('id-0');
  });
});

File: go/receipts_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.

package telemetry

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"testing"
	"time"
)

// resetReceipts resets receipt state and registers cleanup for t.
func resetReceipts(t *testing.T) {
	t.Helper()
	_resetPIIRules()
	ResetReceiptsForTests()
	_resetHealth()
	t.Cleanup(func() {
		_resetPIIRules()
		ResetReceiptsForTests()
		_resetHealth()
	})
}

// mustEnableReceipts fails the test if the sink contract rejects opts.
func mustEnableReceipts(t *testing.T, opts ReceiptOptions) {
	t.Helper()
	if err := EnableReceipts(opts); err != nil {
		t.Fatalf("EnableReceipts(%+v): %v", opts, err)
	}
}

// TestReceiptsDisabledByDefault verifies no receipts are emitted before EnableReceipts is called.
func TestReceiptsDisabledByDefault(t *testing.T) {
	resetReceipts(t)
	payload := map[string]any{"password": "secret123"}
	SanitizePayload(payload, true, 0)
	receipts := GetEmittedReceiptsForTests()
	if len(receipts) != 0 {
		t.Errorf("expected 0 receipts, got %d", len(receipts))
	}
}

// TestReceiptsEmittedWhenEnabled verifies receipts are generated after EnableReceipts.
func TestReceiptsEmittedWhenEnabled(t *testing.T) {
	resetReceipts(t)
	mustEnableReceipts(t, ReceiptOptions{Enabled: true, ServiceName: "test-svc"})
	payload := map[string]any{"password": "secret123"}
	SanitizePayload(payload, true, 0)
	receipts := GetEmittedReceiptsForTests()
	if len(receipts) != 1 {
		t.Fatalf("expected 1 receipt, got %d", len(receipts))
	}
	r := receipts[0]
	if r.FieldPath != "password" {
		t.Errorf("expected field_path 'password', got %q", r.FieldPath)
	}
	if r.Action != "redact" {
		t.Errorf("expected action 'redact', got %q", r.Action)
	}
	if r.ReceiptID == "" {
		t.Error("expected non-empty receipt_id")
	}
}

// TestReceiptOriginalHashIsSHA256 verifies the hash is SHA-256 of the original value.
func TestReceiptOriginalHashIsSHA256(t *testing.T) {
	resetReceipts(t)
	mustEnableReceipts(t, ReceiptOptions{Enabled: true})
	payload := map[string]any{"password": "secret123"}
	SanitizePayload(payload, true, 0)
	receipts := GetEmittedReceiptsForTests()
	if len(receipts) != 1 {
		t.Fatalf("expected 1 receipt, got %d", len(receipts))
	}
	// Canonical JSON, not the %v rendering: the string "1" and the number 1
	// share a display form and must not share a digest.
	sum := sha256.Sum256([]byte(`"secret123"`))
	expected := hex.EncodeToString(sum[:])
	if receipts[0].OriginalHash != expected {
		t.Errorf("hash mismatch: expected %q, got %q", expected, receipts[0].OriginalHash)
	}
}

// TestReceiptHMACWhenKeyProvided verifies HMAC is computed correctly when a key is set.
func TestReceiptHMACWhenKeyProvided(t *testing.T) {
	resetReceipts(t)
	mustEnableReceipts(t, ReceiptOptions{Enabled: true, SigningKey: "test-key"})
	payload := map[string]any{"password": "secret123"}
	SanitizePayload(payload, true, 0)
	receipts := GetEmittedReceiptsForTests()
	if len(receipts) != 1 {
		t.Fatalf("expected 1 receipt, got %d", len(receipts))
	}
	r := receipts[0]
	if r.HMAC == "" {
		t.Error("expected non-empty HMAC")
	}
	payloadStr := fmt.Sprintf("%s|%s|%s|%s|%s",
		r.ReceiptID, r.Timestamp, r.FieldPath, r.Action, r.OriginalHash)
	mac := hmac.New(sha256.New, []byte("test-key"))
	mac.Write([]byte(payloadStr)) //nolint:errcheck
	expected := hex.EncodeToString(mac.Sum(nil))
	if r.HMAC != expected {
		t.Errorf("HMAC mismatch: expected %q, got %q", expected, r.HMAC)
	}
}

// TestReceiptHMACEmptyWhenNoKey verifies HMAC is empty when no signing key is provided.
func TestReceiptHMACEmptyWhenNoKey(t *testing.T) {
	resetReceipts(t)
	mustEnableReceipts(t, ReceiptOptions{Enabled: true})
	payload := map[string]any{"password": "secret123"}
	SanitizePayload(payload, true, 0)
	receipts := GetEmittedReceiptsForTests()
	if len(receipts) != 1 {
		t.Fatalf("expected 1 receipt, got %d", len(receipts))
	}
	if receipts[0].HMAC != "" {
		t.Errorf("expected empty HMAC, got %q", receipts[0].HMAC)
	}
}

// TestReceiptTamperDetection verifies that changing field_path produces a different HMAC.
func TestReceiptTamperDetection(t *testing.T) {
	resetReceipts(t)
	mustEnableReceipts(t, ReceiptOptions{Enabled: true, SigningKey: "test-key"})
	payload := map[string]any{"password": "secret123"}
	SanitizePayload(payload, true, 0)
	receipts := GetEmittedReceiptsForTests()
	if len(receipts) != 1 {
		t.Fatalf("expected 1 receipt, got %d", len(receipts))
	}
	r := receipts[0]
	tamperedPayload := fmt.Sprintf("%s|%s|%s|%s|%s",
		r.ReceiptID, r.Timestamp, "tampered.path", r.Action, r.OriginalHash)
	mac := hmac.New(sha256.New, []byte("test-key"))
	mac.Write([]byte(tamperedPayload)) //nolint:errcheck
	tamperedHMAC := hex.EncodeToString(mac.Sum(nil))
	if r.HMAC == tamperedHMAC {
		t.Error("expected HMAC to differ after tampering with field_path")
	}
}

// TestEnableReceiptsDisabled verifies that EnableReceipts(false,...) unregisters the hook.
func TestEnableReceiptsDisabled(t *testing.T) {
	resetReceipts(t)
	mustEnableReceipts(t, ReceiptOptions{Enabled: true})
	_piiMu.RLock()
	hook := _receiptHook
	_piiMu.RUnlock()
	if hook == nil {
		t.Error("expected hook to be set after EnableReceipts(true)")
	}
	mustEnableReceipts(t, ReceiptOptions{})
	_piiMu.RLock()
	hook = _receiptHook
	_piiMu.RUnlock()
	if hook != nil {
		t.Error("expected hook to be nil after EnableReceipts(false)")
	}
}

// TestReceiptIDIsUUIDFormat verifies receipt_id has UUID format.
func TestReceiptIDIsUUIDFormat(t *testing.T) {
	resetReceipts(t)
	mustEnableReceipts(t, ReceiptOptions{Enabled: true})
	payload := map[string]any{"password": "secret123"}
	SanitizePayload(payload, true, 0)
	receipts := GetEmittedReceiptsForTests()
	if len(receipts) != 1 {
		t.Fatalf("expected 1 receipt, got %d", len(receipts))
	}
	rid := receipts[0].ReceiptID
	if len(rid) != 36 {
		t.Errorf("expected UUID length 36, got %d", len(rid))
	}
	parts := strings.Split(rid, "-")
	if len(parts) != 5 {
		t.Errorf("expected 5 UUID parts, got %d", len(parts))
	}
}

// _enterProductionReceiptMode switches the receipt engine out of test mode for
// the duration of t, so the sink contract can be exercised as a caller sees it.
func _enterProductionReceiptMode(t *testing.T) {
	t.Helper()
	resetReceipts(t)
	_receiptsMu.Lock()
	_receiptsTestMode = false
	_receiptsMu.Unlock()
}

// TestReceiptsProductionMode verifies production receipts reach the configured
// sink and never the test collector.
func TestReceiptsProductionMode(t *testing.T) {
	_enterProductionReceiptMode(t)

	sink := &TestReceiptCollector{}
	mustEnableReceipts(t, ReceiptOptions{Enabled: true, ServiceName: "prod-svc", Sink: sink})
	SanitizePayload(map[string]any{"password": "secret123"}, true, 0)

	delivered := sink.Receipts()
	if len(delivered) != 1 {
		t.Fatalf("expected 1 receipt at the sink, got %d", len(delivered))
	}
	if delivered[0].ServiceName != "prod-svc" {
		t.Errorf("service name: got %q", delivered[0].ServiceName)
	}
	if collected := GetEmittedReceiptsForTests(); len(collected) != 0 {
		t.Errorf("test collector took delivery outside test mode: %d", len(collected))
	}
	if got := GetHealthSnapshot().ReceiptFailures; got != 0 {
		t.Errorf("ReceiptFailures: want 0, got %d", got)
	}
}

// TestEnableReceiptsRequiresSinkInProduction pins the contract that enabling
// receipts without a destination is an error, not a silent discard: a service
// would otherwise sign an audit record per redaction and drop every one.
func TestEnableReceiptsRequiresSinkInProduction(t *testing.T) {
	_enterProductionReceiptMode(t)

	if err := EnableReceipts(ReceiptOptions{Enabled: true, SigningKey: "k"}); !errors.Is(err, ErrMissingReceiptSink) {
		t.Fatalf("want ErrMissingReceiptSink, got %v", err)
	}
	_piiMu.RLock()
	hook := _receiptHook
	_piiMu.RUnlock()
	if hook != nil {
		t.Error("a rejected EnableReceipts must not register the hook")
	}
	// Disabling never needs a sink.
	if err := EnableReceipts(ReceiptOptions{}); err != nil {
		t.Fatalf("disabling receipts: %v", err)
	}
}

type _rejectingSink struct{}

func (_rejectingSink) Emit(RedactionReceipt) bool { return false }

type _panickingSink struct{}

func (_panickingSink) Emit(RedactionReceipt) bool { panic("sink exploded") }

// TestReceiptSinkFailuresCountIntoHealth covers the three ways a receipt fails
// to arrive. All three are counted and none of them logs: the logger is what
// produces redactions, so logging here would be an unbounded cycle.
func TestReceiptSinkFailuresCountIntoHealth(t *testing.T) {
	cases := []struct {
		name string
		sink ReceiptSink
	}{
		{"refusing sink", _rejectingSink{}},
		{"panicking sink", _panickingSink{}},
		{"absent sink", nil},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			resetReceipts(t)
			emitReceipt(RedactionReceipt{ReceiptID: "r"}, tc.sink)
			if got := GetHealthSnapshot().ReceiptFailures; got != 1 {
				t.Fatalf("ReceiptFailures: want 1, got %d", got)
			}
		})
	}
}

// TestTestReceiptCollectorEvictsOldest pins the cap on the one sink this
// library owns. A production sink is never capped.
func TestTestReceiptCollectorEvictsOldest(t *testing.T) {
	collector := &TestReceiptCollector{}
	for i := range TestReceiptCapacity + 2 {
		collector.Emit(RedactionReceipt{ReceiptID: strconv.Itoa(i)})
	}
	got := collector.Receipts()
	if len(got) != TestReceiptCapacity {
		t.Fatalf("retained %d receipts, want %d", len(got), TestReceiptCapacity)
	}
	if got[0].ReceiptID != "2" {
		t.Errorf("oldest retained receipt: got %q, want %q", got[0].ReceiptID, "2")
	}
	if got[len(got)-1].ReceiptID != strconv.Itoa(TestReceiptCapacity+1) {
		t.Errorf("newest retained receipt: got %q", got[len(got)-1].ReceiptID)
	}
	collector.Reset()
	if n := len(collector.Receipts()); n != 0 {
		t.Errorf("Reset left %d receipts", n)
	}
}

// TestReceiptTimestampIsFixedWidthMillis pins the wire spelling. RFC3339Nano
// trims trailing zeros, so the same instant would format to different widths
// depending on its fractional part.
func TestReceiptTimestampIsFixedWidthMillis(t *testing.T) {
	at := time.Date(2026, 8, 4, 12, 34, 56, 789000000, time.UTC)
	if got := _receiptTimestamp(at); got != "2026-08-04T12:34:56.789Z" {
		t.Errorf("got %q", got)
	}
	if got := _receiptTimestamp(time.Date(2026, 8, 4, 12, 34, 56, 0, time.UTC)); got != "2026-08-04T12:34:56.000Z" {
		t.Errorf("zero fraction: got %q", got)
	}
}
